File: src/synapt_eval/report_card/compose.py
"""Compose a ReportCard from upstream eval phase outputs."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

from synapt_eval.report_card.types import (
    CategorySection,
    ReportCard,
    ReportCardFooter,
    ReportCardHeader,
    TrendingEntry,
)
from synapt_eval.runner.orchestration import compute_deltas
from synapt_eval.suggestion_engine.types import Suggestion
from synapt_eval.types import EvalResult


def compose_report_card(
    results: list[EvalResult],
    suggestions: list[Suggestion] | None = None,
    baseline: list[EvalResult] | None = None,
    history: list[EvalResult] | None = None,
    run_id: str | None = None,
    timestamp: str | None = None,
    commit: str | None = None,
    config: dict[str, Any] | None = None,
) -> ReportCard:
    """Build a ReportCard from eval results and upstream outputs."""
    suggestions = suggestions or []

    now = datetime.now(timezone.utc)
    run_id = run_id or now.strftime("%Y%m%dT%H%M%SZ")
    timestamp = timestamp or now.isoformat()

    severity_counts = _count_severities(suggestions)
    fixture_count = sum(r.metrics.n for r in results)

    header = ReportCardHeader(
        run_id=run_id,
        timestamp=timestamp,
        fixture_count=fixture_count,
        category_count=len(results),
        severity_counts=severity_counts,
        commit=commit,
        config=config or {},
    )

    suggestion_by_category: dict[str, list[Suggestion]] = {}
    for s in suggestions:
        cat = s.category or "_global"
        suggestion_by_category.setdefault(cat, []).append(s)

    sections = [
        CategorySection(
            category=r.category,
            metrics=r.metrics,
            suggestions=suggestion_by_category.get(r.category, []),
            fixture_count=r.metrics.n,
        )
        for r in results
    ]

    deltas = []
    regression_summary = None
    if baseline:
        deltas = compute_deltas(results, baseline)
        regressions = [d for d in deltas if d.regression]
        if regressions:
            regression_summary = f"{len(regressions)} regression(s) detected"

    has_errors = severity_counts.get("error", 0) > 0
    has_regressions = regression_summary is not None
    passed = not has_errors and not has_regressions

    footer = ReportCardFooter(
        passed=passed,
        total_suggestions=len(suggestions),
        error_count=severity_counts.get("error", 0),
        warning_count=severity_counts.get("warning", 0),
        info_count=severity_counts.get("info", 0),
        regression_summary=regression_summary,
        deltas=deltas,
    )

    trending = _build_trending(history) if history else []

    return ReportCard(
        header=header,
        sections=sections,
        suggestions=suggestions,
        footer=footer,
        trending=trending,
    )


def _count_severities(suggestions: list[Suggestion]) -> dict[str, int]:
    counts: dict[str, int] = {"error": 0, "warning": 0, "info": 0}
    for s in suggestions:
        level = s.severity.level
        if level in counts:
            counts[level] += 1
        elif s.severity.weight >= 0.75:
            counts["error"] += 1
        elif s.severity.weight >= 0.5:
            counts["warning"] += 1
        else:
            counts["info"] += 1
    return counts


def _build_trending(
    history: list[EvalResult],
) -> list[TrendingEntry]:
    entries: list[TrendingEntry] = []
    for r in history:
        metrics: dict[str, float] = {
            "p_at_5": r.metrics.p_at_5,
            "r_at_10": r.metrics.r_at_10,
        }
        if r.metrics.tau is not None:
            metrics["tau"] = r.metrics.tau
        entries.append(
            TrendingEntry(
                run_id="historical",
                category=r.category,
                metrics=metrics,
            )
        )
    return entries

File: ts/src/report_card/compose.ts
import type { EvalResult } from "../types.ts";
import type { Suggestion } from "../suggestion_engine/types.ts";
import type {
  CategorySection,
  Delta,
  ReportCard,
  ReportCardFooter,
  ReportCardHeader,
  TrendingEntry,
} from "./types.ts";

export interface ComposeOptions {
  results: EvalResult[];
  suggestions?: Suggestion[];
  baseline?: EvalResult[];
  history?: EvalResult[];
  runId?: string;
  timestamp?: string;
  commit?: string;
  config?: Record<string, unknown>;
}

export function composeReportCard(options: ComposeOptions): ReportCard {
  const suggestions = options.suggestions ?? [];
  const now = new Date();
  const runId = options.runId ?? now.toISOString().replace(/[-:]/g, "").slice(0, 15) + "Z";
  const timestamp = options.timestamp ?? now.toISOString();

  const severityCounts = countSeverities(suggestions);
  const fixtureCount = options.results.reduce((sum, r) => sum + r.metrics.n, 0);

  const header: ReportCardHeader = {
    runId,
    timestamp,
    fixtureCount,
    categoryCount: options.results.length,
    severityCounts,
    commit: options.commit,
    config: options.config ?? {},
  };

  const suggestionByCategory = new Map<string, Suggestion[]>();
  for (const s of suggestions) {
    const cat = s.category ?? "_global";
    const list = suggestionByCategory.get(cat) ?? [];
    list.push(s);
    suggestionByCategory.set(cat, list);
  }

  const sections: CategorySection[] = options.results.map((r) => ({
    category: r.category,
    metrics: r.metrics,
    suggestions: suggestionByCategory.get(r.category) ?? [],
    fixtureCount: r.metrics.n,
  }));

  let deltas: Delta[] = [];
  let regressionSummary: string | undefined;
  if (options.baseline) {
    deltas = computeDeltas(options.results, options.baseline);
    const regressions = deltas.filter((d) => d.regression);
    if (regressions.length > 0) {
      regressionSummary = `${regressions.length} regression(s) detected`;
    }
  }

  const hasErrors = (severityCounts["error"] ?? 0) > 0;
  const hasRegressions = regressionSummary !== undefined;
  const passed = !hasErrors && !hasRegressions;

  const footer: ReportCardFooter = {
    passed,
    totalSuggestions: suggestions.length,
    errorCount: severityCounts["error"] ?? 0,
    warningCount: severityCounts["warning"] ?? 0,
    infoCount: severityCounts["info"] ?? 0,
    regressionSummary,
    deltas,
  };

  const trending = options.history ? buildTrending(options.history) : [];

  return { header, sections, suggestions, footer, trending };
}

function countSeverities(suggestions: Suggestion[]): Record<string, number> {
  const counts: Record<string, number> = { error: 0, warning: 0, info: 0 };
  for (const s of suggestions) {
    const level = s.severity.level;
    if (level in counts) {
      counts[level]++;
    } else if (s.severity.weight >= 0.75) {
      counts["error"]++;
    } else if (s.severity.weight >= 0.5) {
      counts["warning"]++;
    } else {
      counts["info"]++;
    }
  }
  return counts;
}

function computeDeltas(
  results: EvalResult[],
  baseline: EvalResult[],
  threshold = 0.05,
): Delta[] {
  const deltas: Delta[] = [];
  for (const result of results) {
    const base = baseline.find((b) => b.category === result.category);
    if (!base) continue;

    const metrics: Array<{ key: "pAt5" | "rAt10"; name: string }> = [
      { key: "pAt5", name: "p_at_5" },
      { key: "rAt10", name: "r_at_10" },
    ];

    for (const { key, name } of metrics) {
      const current = result.metrics[key];
      const baseVal = base.metrics[key];
      const delta = current - baseVal;
      deltas.push({
        category: result.category,
        metric: name,
        baseline: baseVal,
        current,
        delta,
        regression: delta < -threshold,
      });
    }
  }
  return deltas;
}

function buildTrending(history: EvalResult[]): TrendingEntry[] {
  return history.map((r) => {
    const metrics: Record<string, number> = {
      p_at_5: r.metrics.pAt5,
      r_at_10: r.metrics.rAt10,
    };
    if (r.metrics.tau != null) {
      metrics["tau"] = r.metrics.tau;
    }
    return { runId: "historical", category: r.category, metrics };
  });
}
